fix message ordering for wire_api = "chat"

drifkin opened the pull request:
In the chat/completions path, whenever a model sent assistant text ahead of a tool call (with or without reasoning), the text was buffered, but wasn't flushed until after the tool call, so the call appeared before the assistant reply. And in the aggregated view the buffered message could be emitted twice.

The gpt-oss models don't exhibit this behavior since they only interleave tool calling with reasoning (not any final output), but other models will do this (e.g., `gpt-5` with `wire_api` set to `"chat"`)

Added tests that fail with what's currently on main.

etraut-openai commented:
Thanks for the contribution, and apologies for the slow response. We've received many PRs, and we don't have the bandwidth on the codex team to review all of them.

It looks like this PR attempts to address a bug, but I don't see any bug report. If you have identified a bug, please start by filing a bug report that describes the problem and its impact so we can triage and prioritize. I took a quick look at the contents of this PR, and it appears to be a substantive change. We're more likely to review and accept PRs that are targeted and surgical.

etraut-openai commented:
I think I was too hasty in closing this PR. Reopening.

etraut-openai commented:
@drifkin, sorry that this PR has languished for so long. We fell behind on PRs, and this one slipped through the cracks. I've asked a member of the codex team to take a look at this, and it's in his queue. Assuming that you're still interested in pursuing this PR, could you please resolve the merge conflicts and get it back into a state where it can be reviewed? Thanks!

drifkin commented:
hi @etraut-openai, thanks for the update! I'll try to get it rebased this weekend.

drifkin committed:
fix message ordering for wire_api = "chat"

In the chat/completions path, whenever a model sent assistant text ahead
of a tool call (with or without reasoning), the text was buffered, but
wasn't flushed until after the tool call, so the call appeared before
the (actually earlier) assistant reply. And in the aggregated view the
buffered message could be emitted twice.

The gpt-oss models don't exhibit this behavior since they only
interleave tool calling with reasoning (not any final output), but other
models will do this (e.g., `gpt-5` with `wire_api` set to `"chat"`)

Added tests that fail with what's currently on main.

etraut-openai committed:
Merge branch 'main' into chat-completion-bug-fix

etraut-openai commented:
@drifkin, there are lint (clippy) failures that are still blocking CI from passing.

drifkin committed:
fix lint

etraut-openai committed:
Merge branch 'main' into chat-completion-bug-fix

drifkin commented:
> @drifkin, thanks for fixing the clippy issues. The tests are also failing. Could you address those failures? Thanks!

will take a look today or tomorrow, thanks for re-running!

drifkin committed:
fix tests

drifkin commented:
hmm looks like tests are still failing in CI, I'll try to figure out why they were passing locally but failing in CI

drifkin committed:
fix more tests

drifkin committed:
Merge branch 'main' into chat-completion-bug-fix

pakrym-oai commented on codex-rs/app-server/tests/suite/codex_message_processor_flow.rs:
```diff
@@ -158,6 +162,53 @@ async fn test_codex_jsonrpc_conversation_flow() -> Result<()> {
     Ok(())
 }
 
+async fn wait_for_task_complete_handling_exec_approval(
```
is this test change related?

pakrym-oai commented on codex-rs/core/tests/chat_completions_payload.rs:
```diff
@@ -305,15 +305,30 @@ async fn skips_empty_reasoning_segments() {
     skip_if_no_network!();
 
-    let body = run_request(vec![assistant_message("dup"), assistant_message("dup")]).await;
+    let body = run_request(vec![
```
should we keep the existing test?

etraut-openai commented:
@drifkin, please take a look at the code review feedback from @pakrym-oai when you have time. I think this PR is close to ready, and I'd love to push it over the finish line.

pakrym-oai commented on codex-rs/core/tests/chat_completions_sse.rs:
```diff
@@ -381,6 +392,238 @@ async fn streams_reasoning_before_tool_call() {
+        other => panic!("expected reasoning item added, got {other:?}"),
+    }
+
+    match &events[1] {
```
we don't expect items to overlap in agent code. One item should be done before the next gets added.

dkappler commented:
> Thank you for the fix @dkappler!
> 
> My main question is whether we need to change both formatting and parsing sides?
> 
> In theory if messages are ordered correctly on the parsing side you'll get them back in the same order on the next request.

I think that's s typo, I did not contribute to this.

drifkin commented:
thanks @etraut-openai and @pakrym-oai! I'm traveling at the moment, so it might take me a little bit longer. But I'll try to make some changes soon, especially to see if I run into any trouble simplifying it to only parser changes.

drifkin committed:
simplify chat payload tool-call handling and rely on SSE ordering

drifkin commented:
@pakrym-oai: I've simplified the approach with 2100afa, could you see if that more aligns with what you were thinking? If all looks good, then I can update the branch again to be cleanly mergeable
